File: internal/envelope/envelope.go
// Package envelope implements the client-side crypto workflow for secrt.ca v1.
//
// It provides Seal (encrypt), Open (decrypt), key derivation, claim token/hash
// computation, URL fragment parsing, and TTL parsing. All operations are pure
// (no I/O, no network) and safe for concurrent use.
package envelope

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/url"
	"strings"

	"golang.org/x/crypto/hkdf"
	"golang.org/x/crypto/pbkdf2"
)

// Crypto constants from spec/v1/envelope.md.
const (
	URLKeyLen     = 32
	PassKeyLen    = 32
	HKDFLen       = 32
	GCMNonceLen   = 12
	HKDFSaltLen   = 32
	KDFSaltLen    = 16
	AAD           = "secrt.ca/envelope/v1"
	HKDFInfoEnc   = "secret:v1:enc"
	HKDFInfoClaim = "secret:v1:claim"
	Suite         = "v1-pbkdf2-hkdf-aes256gcm"

	DefaultPBKDF2Iterations = 600000
	MinPBKDF2Iterations     = 300000
)

// Errors returned by envelope operations.
var (
	ErrEmptyPlaintext   = errors.New("plaintext must not be empty")
	ErrInvalidEnvelope  = errors.New("invalid envelope")
	ErrDecryptionFailed = errors.New("decryption failed")
	ErrInvalidFragment  = errors.New("invalid URL fragment")
	ErrInvalidURLKey    = errors.New("url_key must be 32 bytes")
)

// Envelope is the JSON structure stored on the server.
type Envelope struct {
	V     int               `json:"v"`
	Suite string            `json:"suite"`
	Enc   EncBlock          `json:"enc"`
	KDF   json.RawMessage   `json:"kdf"`
	HKDF  HKDFBlock         `json:"hkdf"`
	Hint  map[string]string `json:"hint,omitempty"`
}

// EncBlock holds the AES-GCM ciphertext.
type EncBlock struct {
	Alg        string `json:"alg"`
	Nonce      string `json:"nonce"`
	Ciphertext string `json:"ciphertext"`
}

// HKDFBlock holds the HKDF parameters.
type HKDFBlock struct {
	Hash      string `json:"hash"`
	Salt      string `json:"salt"`
	EncInfo   string `json:"enc_info"`
	ClaimInfo string `json:"claim_info"`
	Length    int    `json:"length"`
}

// KDFNone is the KDF block when no passphrase is used.
type KDFNone struct {
	Name string `json:"name"`
}

// KDFPBKDF2 is the KDF block when a passphrase is used.
type KDFPBKDF2 struct {
	Name       string `json:"name"`
	Salt       string `json:"salt"`
	Iterations int    `json:"iterations"`
	Length     int    `json:"length"`
}

// kdfParsed is the internal representation after parsing KDF JSON.
type kdfParsed struct {
	Name       string
	Salt       []byte
	Iterations int
}

// SealParams holds inputs for creating an encrypted envelope.
type SealParams struct {
	Plaintext  []byte
	Passphrase string            // empty = no passphrase
	Rand       io.Reader         // nil = crypto/rand.Reader
	Hint       map[string]string // optional hint metadata
	Iterations int               // PBKDF2 iterations; 0 = DefaultPBKDF2Iterations
}

// SealResult holds outputs from creating an encrypted envelope.
type SealResult struct {
	Envelope   json.RawMessage // envelope JSON
	URLKey     []byte          // 32-byte URL key (for share link fragment)
	ClaimToken []byte          // 32-byte claim token
	ClaimHash  string          // base64url(SHA-256(claim_token))
}

// OpenParams holds inputs for decrypting an envelope.
type OpenParams struct {
	Envelope   json.RawMessage // envelope JSON from server
	URLKey     []byte          // 32-byte URL key from share link
	Passphrase string          // empty = no passphrase
}

func rng(r io.Reader) io.Reader {
	if r != nil {
		return r
	}
	return rand.Reader
}

func readRand(r io.Reader, n int) ([]byte, error) {
	buf := make([]byte, n)
	if _, err := io.ReadFull(r, buf); err != nil {
		return nil, fmt.Errorf("read random bytes: %w", err)
	}
	return buf, nil
}

func b64Encode(data []byte) string {
	return base64.RawURLEncoding.EncodeToString(data)
}

func b64Decode(s string) ([]byte, error) {
	return base64.RawURLEncoding.DecodeString(s)
}

func deriveHKDF(ikm, salt []byte, info string, length int) ([]byte, error) {
	r := hkdf.New(sha256.New, ikm, salt, []byte(info))
	out := make([]byte, length)
	if _, err := io.ReadFull(r, out); err != nil {
		return nil, fmt.Errorf("HKDF derive: %w", err)
	}
	return out, nil
}

// DeriveClaimToken derives a claim token from url_key alone.
// claim_token = HKDF-SHA-256(url_key, nil, "secret:v1:claim", 32)
func DeriveClaimToken(urlKey []byte) ([]byte, error) {
	if len(urlKey) != URLKeyLen {
		return nil, ErrInvalidURLKey
	}
	return deriveHKDF(urlKey, nil, HKDFInfoClaim, HKDFLen)
}

// ComputeClaimHash returns base64url(SHA-256(claim_token)).
func ComputeClaimHash(claimToken []byte) string {
	h := sha256.Sum256(claimToken)
	return b64Encode(h[:])
}

// Seal creates an encrypted envelope from plaintext.
func Seal(p SealParams) (SealResult, error) {
	if len(p.Plaintext) == 0 {
		return SealResult{}, ErrEmptyPlaintext
	}

	r := rng(p.Rand)

	// 1. Generate url_key
	urlKey, err := readRand(r, URLKeyLen)
	if err != nil {
		return SealResult{}, err
	}

	// 2. Build KDF + compute IKM
	var ikm []byte
	var kdfJSON json.RawMessage

	if p.Passphrase == "" {
		ikm = urlKey
		kdfJSON, err = json.Marshal(KDFNone{Name: "none"})
		if err != nil {
			return SealResult{}, err
		}
	} else {
		kdfSalt, err := readRand(r, KDFSaltLen)
		if err != nil {
			return SealResult{}, err
		}
		iterations := p.Iterations
		if iterations == 0 {
			iterations = DefaultPBKDF2Iterations
		}
		passKey := pbkdf2.Key([]byte(p.Passphrase), kdfSalt, iterations, PassKeyLen, sha256.New)
		h := sha256.New()
		h.Write(urlKey)
		h.Write(passKey)
		ikm = h.Sum(nil)
		kdfJSON, err = json.Marshal(KDFPBKDF2{
			Name:       "PBKDF2-SHA256",
			Salt:       b64Encode(kdfSalt),
			Iterations: iterations,
			Length:     PassKeyLen,
		})
		if err != nil {
			return SealResult{}, err
		}
	}

	// 3. Generate HKDF salt
	hkdfSalt, err := readRand(r, HKDFSaltLen)
	if err != nil {
		return SealResult{}, err
	}

	// 4. Derive enc_key
	encKey, err := deriveHKDF(ikm, hkdfSalt, HKDFInfoEnc, HKDFLen)
	if err != nil {
		return SealResult{}, err
	}

	// 5. Derive claim_token (from url_key alone)
	claimToken, err := DeriveClaimToken(urlKey)
	if err != nil {
		return SealResult{}, err
	}

	// 6. Generate nonce
	nonce, err := readRand(r, GCMNonceLen)
	if err != nil {
		return SealResult{}, err
	}

	// 7. Encrypt
	block, err := aes.NewCipher(encKey)
	if err != nil {
		return SealResult{}, fmt.Errorf("AES cipher: %w", err)
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return SealResult{}, fmt.Errorf("GCM: %w", err)
	}
	ciphertext := gcm.Seal(nil, nonce, p.Plaintext, []byte(AAD))

	// 8. Build envelope
	env := Envelope{
		V:     1,
		Suite: Suite,
		Enc: EncBlock{
			Alg:        "A256GCM",
			Nonce:      b64Encode(nonce),
			Ciphertext: b64Encode(ciphertext),
		},
		KDF: kdfJSON,
		HKDF: HKDFBlock{
			Hash:      "SHA-256",
			Salt:      b64Encode(hkdfSalt),
			EncInfo:   HKDFInfoEnc,
			ClaimInfo: HKDFInfoClaim,
			Length:    HKDFLen,
		},
		Hint: p.Hint,
	}

	envJSON, err := json.Marshal(env)
	if err != nil {
		return SealResult{}, fmt.Errorf("marshal envelope: %w", err)
	}

	return SealResult{
		Envelope:   envJSON,
		URLKey:     urlKey,
		ClaimToken: claimToken,
		ClaimHash:  ComputeClaimHash(claimToken),
	}, nil
}

// Open decrypts an envelope, returning plaintext.
func Open(p OpenParams) ([]byte, error) {
	if len(p.URLKey) != URLKeyLen {
		return nil, ErrInvalidURLKey
	}

	// Parse envelope
	var env Envelope
	if err := json.Unmarshal(p.Envelope, &env); err != nil {
		return nil, fmt.Errorf("%w: %w", ErrInvalidEnvelope, err)
	}

	if err := validateEnvelope(&env); err != nil {
		return nil, err
	}

	// Parse KDF
	kdf, err := parseKDF(env.KDF)
	if err != nil {
		return nil, err
	}

	// Compute IKM
	var ikm []byte
	if kdf.Name == "none" {
		ikm = p.URLKey
	} else {
		passKey := pbkdf2.Key([]byte(p.Passphrase), kdf.Salt, kdf.Iterations, PassKeyLen, sha256.New)
		h := sha256.New()
		h.Write(p.URLKey)
		h.Write(passKey)
		ikm = h.Sum(nil)
	}

	// Derive enc_key
	hkdfSalt, err := b64Decode(env.HKDF.Salt)
	if err != nil {
		return nil, fmt.Errorf("%w: invalid hkdf.salt", ErrInvalidEnvelope)
	}
	encKey, err := deriveHKDF(ikm, hkdfSalt, HKDFInfoEnc, HKDFLen)
	if err != nil {
		return nil, err
	}

	// Decode nonce and ciphertext
	nonce, err := b64Decode(env.Enc.Nonce)
	if err != nil {
		return nil, fmt.Errorf("%w: invalid nonce", ErrInvalidEnvelope)
	}
	ciphertext, err := b64Decode(env.Enc.Ciphertext)
	if err != nil {
		return nil, fmt.Errorf("%w: invalid ciphertext", ErrInvalidEnvelope)
	}

	// Decrypt
	block, err := aes.NewCipher(encKey)
	if err != nil {
		return nil, fmt.Errorf("AES cipher: %w", err)
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, fmt.Errorf("GCM: %w", err)
	}
	plaintext, err := gcm.Open(nil, nonce, ciphertext, []byte(AAD))
	if err != nil {
		return nil, ErrDecryptionFailed
	}

	return plaintext, nil
}

func validateEnvelope(env *Envelope) error {
	if env.V != 1 {
		return fmt.Errorf("%w: unsupported version %d", ErrInvalidEnvelope, env.V)
	}
	if env.Suite != Suite {
		return fmt.Errorf("%w: unsupported suite %q", ErrInvalidEnvelope, env.Suite)
	}
	if env.Enc.Alg != "A256GCM" {
		return fmt.Errorf("%w: unsupported enc.alg %q", ErrInvalidEnvelope, env.Enc.Alg)
	}

	nonce, err := b64Decode(env.Enc.Nonce)
	if err != nil || len(nonce) != GCMNonceLen {
		return fmt.Errorf("%w: nonce must be %d bytes", ErrInvalidEnvelope, GCMNonceLen)
	}

	ct, err := b64Decode(env.Enc.Ciphertext)
	if err != nil || len(ct) < 16 {
		return fmt.Errorf("%w: ciphertext too short (need at least GCM tag)", ErrInvalidEnvelope)
	}

	if env.HKDF.Hash != "SHA-256" {
		return fmt.Errorf("%w: unsupported hkdf.hash %q", ErrInvalidEnvelope, env.HKDF.Hash)
	}
	hkdfSalt, err := b64Decode(env.HKDF.Salt)
	if err != nil || len(hkdfSalt) != HKDFSaltLen {
		return fmt.Errorf("%w: hkdf.salt must be %d bytes", ErrInvalidEnvelope, HKDFSaltLen)
	}
	if env.HKDF.EncInfo != HKDFInfoEnc {
		return fmt.Errorf("%w: invalid hkdf.enc_info", ErrInvalidEnvelope)
	}
	if env.HKDF.ClaimInfo != HKDFInfoClaim {
		return fmt.Errorf("%w: invalid hkdf.claim_info", ErrInvalidEnvelope)
	}
	if env.HKDF.Length != HKDFLen {
		return fmt.Errorf("%w: hkdf.length must be %d", ErrInvalidEnvelope, HKDFLen)
	}

	return nil
}

func parseKDF(raw json.RawMessage) (kdfParsed, error) {
	var probe struct {
		Name string `json:"name"`
	}
	if err := json.Unmarshal(raw, &probe); err != nil {
		return kdfParsed{}, fmt.Errorf("%w: invalid kdf", ErrInvalidEnvelope)
	}

	switch probe.Name {
	case "none":
		return kdfParsed{Name: "none"}, nil
	case "PBKDF2-SHA256":
		var k KDFPBKDF2
		if err := json.Unmarshal(raw, &k); err != nil {
			return kdfParsed{}, fmt.Errorf("%w: invalid kdf", ErrInvalidEnvelope)
		}
		salt, err := b64Decode(k.Salt)
		if err != nil || len(salt) < KDFSaltLen {
			return kdfParsed{}, fmt.Errorf("%w: kdf.salt must be at least %d bytes", ErrInvalidEnvelope, KDFSaltLen)
		}
		if k.Iterations < MinPBKDF2Iterations {
			return kdfParsed{}, fmt.Errorf("%w: kdf.iterations must be >= %d", ErrInvalidEnvelope, MinPBKDF2Iterations)
		}
		if k.Length != PassKeyLen {
			return kdfParsed{}, fmt.Errorf("%w: kdf.length must be %d", ErrInvalidEnvelope, PassKeyLen)
		}
		return kdfParsed{
			Name:       "PBKDF2-SHA256",
			Salt:       salt,
			Iterations: k.Iterations,
		}, nil
	default:
		return kdfParsed{}, fmt.Errorf("%w: unsupported kdf.name %q", ErrInvalidEnvelope, probe.Name)
	}
}

// ParseShareURL extracts id and url_key from a share URL with fragment.
// Accepts formats:
//   - https://host/s/<id>#<url_key_b64>
//   - <id>#<url_key_b64> (bare ID with fragment)
func ParseShareURL(rawURL string) (id string, urlKey []byte, err error) {
	// Try parsing as full URL
	u, parseErr := url.Parse(rawURL)
	if parseErr == nil && u.Scheme != "" {
		// Full URL: extract ID from path
		path := strings.TrimPrefix(u.Path, "/s/")
		if path == u.Path || path == "" {
			return "", nil, fmt.Errorf("%w: expected /s/<id> path", ErrInvalidFragment)
		}
		id = path
	} else {
		// Bare format: id#fragment
		parts := strings.SplitN(rawURL, "#", 2)
		if len(parts) != 2 || parts[0] == "" {
			return "", nil, fmt.Errorf("%w: missing fragment", ErrInvalidFragment)
		}
		id = parts[0]
		if u == nil {
			u = &url.URL{}
		}
		u.Fragment = parts[1]
	}

	// Parse fragment
	frag := u.Fragment
	if frag == "" {
		// url.Parse may strip fragment; try manual extraction
		if idx := strings.Index(rawURL, "#"); idx >= 0 {
			frag = rawURL[idx+1:]
		}
	}

	keyB64 := frag
	urlKey, err = b64Decode(keyB64)
	if err != nil {
		return "", nil, fmt.Errorf("%w: invalid url_key encoding", ErrInvalidFragment)
	}
	if len(urlKey) != URLKeyLen {
		return "", nil, fmt.Errorf("%w: url_key must be %d bytes, got %d", ErrInvalidFragment, URLKeyLen, len(urlKey))
	}

	return id, urlKey, nil
}

// FormatShareLink builds a share URL with fragment.
func FormatShareLink(shareURL string, urlKey []byte) string {
	return shareURL + "#" + b64Encode(urlKey)
}

File: web/src/crypto/envelope.ts
import type {
  PayloadMeta,
  EnvelopeJson,
  SealResult,
  OpenResult,
  KdfArgon2id,
} from '../types';
import { isTauri } from '../lib/config';
import { nativeSeal, nativeOpen, nativeDeriveClaimToken } from './native';
import {
  AAD,
  HKDF_INFO_ENC,
  HKDF_INFO_CLAIM,
  CLAIM_SALT_LABEL,
  SUITE,
  URL_KEY_LEN,
  PASS_KEY_LEN,
  HKDF_LEN,
  GCM_NONCE_LEN,
  HKDF_SALT_LEN,
  KDF_SALT_LEN,
  ARGON2_VERSION,
  ARGON2_M_COST_DEFAULT,
  ARGON2_T_COST_DEFAULT,
  ARGON2_P_COST_DEFAULT,
  ARGON2_M_COST_MIN,
  ARGON2_M_COST_MAX,
  ARGON2_T_COST_MIN,
  ARGON2_T_COST_MAX,
  ARGON2_P_COST_MIN,
  ARGON2_P_COST_MAX,
  ARGON2_M_COST_T_COST_PRODUCT_MAX,
} from './constants';
import {
  base64urlEncode,
  base64urlDecode,
  utf8Encode,
  concatBytes,
} from './encoding';
import { buildFrame, parseFrame } from './frame';
import { deriveArgon2id, preloadArgon2id } from './argon2id';

// TS 5.8 makes Uint8Array generic; Web Crypto expects BufferSource.
// This helper avoids casting at every call site.
const buf = (a: Uint8Array): ArrayBuffer => {
  const b = new ArrayBuffer(a.byteLength);
  new Uint8Array(b).set(a);
  return b;
};

interface SealOptions {
  passphrase?: string;
  rng?: (buf: Uint8Array) => void;
  compress?: (data: Uint8Array) => Uint8Array;
  /** Pre-built frame bytes (already compressed). Skips buildFrame() when provided. */
  prebuiltFrame?: Uint8Array;
}

/**
 * Proactively load Argon2id WASM for passphrase flows.
 * Safe to call multiple times.
 */
export async function preloadPassphraseKdf(): Promise<void> {
  if (isTauri()) return; // Native argon2id needs no WASM preloading
  await preloadArgon2id();
}

/**
 * Seal plaintext content into an encrypted envelope.
 * Returns the envelope JSON, url_key, and claim_hash.
 */
async function webSeal(
  content: Uint8Array,
  meta: PayloadMeta,
  options?: SealOptions,
): Promise<SealResult> {
  const rng = options?.rng ?? cryptoRng;

  // 1. Generate url_key
  const urlKey = new Uint8Array(URL_KEY_LEN);
  rng(urlKey);

  // 2. Build KDF + compute IKM
  let ikm: Uint8Array;
  let kdf: EnvelopeJson['kdf'];

  if (options?.passphrase) {
    const kdfSalt = new Uint8Array(KDF_SALT_LEN);
    rng(kdfSalt);

    const passKey = await deriveArgon2id(
      options.passphrase,
      kdfSalt,
      ARGON2_M_COST_DEFAULT,
      ARGON2_T_COST_DEFAULT,
      ARGON2_P_COST_DEFAULT,
      PASS_KEY_LEN,
    );

    ikm = new Uint8Array(
      await crypto.subtle.digest('SHA-256', buf(concatBytes(urlKey, passKey))),
    );

    kdf = {
      name: 'argon2id',
      version: ARGON2_VERSION,
      salt: base64urlEncode(kdfSalt),
      m_cost: ARGON2_M_COST_DEFAULT,
      t_cost: ARGON2_T_COST_DEFAULT,
      p_cost: ARGON2_P_COST_DEFAULT,
      length: PASS_KEY_LEN,
    };
  } else {
    ikm = urlKey;
    kdf = { name: 'none' };
  }

  // 3. Generate HKDF salt
  const hkdfSalt = new Uint8Array(HKDF_SALT_LEN);
  rng(hkdfSalt);

  // 4. Derive enc_key
  const encKey = await hkdfDerive(ikm, hkdfSalt, HKDF_INFO_ENC, HKDF_LEN);

  // 5. Derive claim_token + claim_hash
  const claimHash = await deriveClaimHash(urlKey);

  // 6. Build framed payload (reuse pre-built frame if provided)
  const frameBytes =
    options?.prebuiltFrame ?? buildFrame(meta, content, options?.compress);

  // 7. Generate nonce
  const nonce = new Uint8Array(GCM_NONCE_LEN);
  rng(nonce);

  // 8. Encrypt
  const aadBytes = utf8Encode(AAD);
  const cryptoKey = await crypto.subtle.importKey(
    'raw',
    buf(encKey),
    'AES-GCM',
    false,
    ['encrypt'],
  );
  const ciphertext = new Uint8Array(
    await crypto.subtle.encrypt(
      { name: 'AES-GCM', iv: buf(nonce), additionalData: buf(aadBytes) },
      cryptoKey,
      buf(frameBytes),
    ),
  );

  // 9. Build envelope
  const envelope: EnvelopeJson = {
    v: 1,
    suite: SUITE,
    enc: {
      alg: 'A256GCM',
      nonce: base64urlEncode(nonce),
      ciphertext: base64urlEncode(ciphertext),
    },
    kdf,
    hkdf: {
      hash: 'SHA-256',
      salt: base64urlEncode(hkdfSalt),
      enc_info: HKDF_INFO_ENC,
      claim_info: HKDF_INFO_CLAIM,
      length: 32,
    },
  };

  return { envelope, urlKey, claimHash };
}

/**
 * Open (decrypt) an envelope using the url_key and optional passphrase.
 */
async function webOpen(
  envelope: EnvelopeJson,
  urlKey: Uint8Array,
  passphrase?: string,
): Promise<OpenResult> {
  // Validate envelope structure
  if (urlKey.length !== URL_KEY_LEN)
    throw new Error(`invalid url_key length: expected ${URL_KEY_LEN} bytes`);
  if (envelope.v !== 1)
    throw new Error(`unsupported envelope version: ${envelope.v}`);
  if (envelope.suite !== SUITE)
    throw new Error(`unsupported suite: ${envelope.suite}`);
  if (envelope.enc.alg !== 'A256GCM')
    throw new Error(`unsupported alg: ${envelope.enc.alg}`);

  const nonce = base64urlDecode(envelope.enc.nonce);
  if (nonce.length !== GCM_NONCE_LEN) throw new Error('invalid nonce length');

  const ciphertext = base64urlDecode(envelope.enc.ciphertext);
  if (ciphertext.length < 16) throw new Error('ciphertext too short');

  if (envelope.hkdf.hash !== 'SHA-256')
    throw new Error(`unsupported hkdf.hash: ${envelope.hkdf.hash}`);
  if (envelope.hkdf.enc_info !== HKDF_INFO_ENC)
    throw new Error('invalid hkdf enc_info');
  if (envelope.hkdf.claim_info !== HKDF_INFO_CLAIM)
    throw new Error('invalid hkdf claim_info');
  if (envelope.hkdf.length !== HKDF_LEN)
    throw new Error(`invalid hkdf length: ${envelope.hkdf.length}`);

  const hkdfSalt = base64urlDecode(envelope.hkdf.salt);
  if (hkdfSalt.length !== HKDF_SALT_LEN)
    throw new Error('invalid hkdf salt length');

  // Recompute IKM
  let ikm: Uint8Array;
  const kdfName = envelope.kdf.name;
  if (kdfName === 'argon2id') {
    if (!passphrase) throw new Error('passphrase required');

    const kdfBlock = envelope.kdf as KdfArgon2id;
    const kdfSalt = parseAndValidateArgon2idKdf(kdfBlock);
    const passKey = await deriveArgon2id(
      passphrase,
      kdfSalt,
      kdfBlock.m_cost,
      kdfBlock.t_cost,
      kdfBlock.p_cost,
      kdfBlock.length,
    );

    ikm = new Uint8Array(
      await crypto.subtle.digest('SHA-256', buf(concatBytes(urlKey, passKey))),
    );
  } else if (kdfName === 'none') {
    assertKdfNoneHasNoExtraFields(envelope.kdf);
    ikm = urlKey;
  } else {
    throw new Error(`unsupported kdf: ${kdfName}`);
  }

  // Derive enc_key
  const encKey = await hkdfDerive(ikm, hkdfSalt, HKDF_INFO_ENC, HKDF_LEN);

  // Decrypt
  const aadBytes = utf8Encode(AAD);
  const cryptoKey = await crypto.subtle.importKey(
    'raw',
    buf(encKey),
    'AES-GCM',
    false,
    ['decrypt'],
  );
  const plaintext = new Uint8Array(
    await crypto.subtle.decrypt(
      { name: 'AES-GCM', iv: buf(nonce), additionalData: buf(aadBytes) },
      cryptoKey,
      buf(ciphertext),
    ),
  );

  // Parse frame
  const { meta, body } = await parseFrame(plaintext);

  return { content: body, meta };
}

/**
 * Derive the claim_hash from a url_key.
 * claim_hash = base64url(SHA-256(claim_token))
 */
export async function deriveClaimHash(urlKey: Uint8Array): Promise<string> {
  const token = await deriveClaimToken(urlKey);
  const hash = new Uint8Array(
    await crypto.subtle.digest('SHA-256', buf(token)),
  );
  return base64urlEncode(hash);
}

/**
 * Derive the claim_token from a url_key.
 * claim_token = HKDF-SHA-256(url_key, CLAIM_SALT, HKDF_INFO_CLAIM, 32)
 */
async function webDeriveClaimToken(
  urlKey: Uint8Array,
): Promise<Uint8Array> {
  const claimSalt = new Uint8Array(
    await crypto.subtle.digest('SHA-256', buf(utf8Encode(CLAIM_SALT_LABEL))),
  );
  return hkdfDerive(urlKey, claimSalt, HKDF_INFO_CLAIM, HKDF_LEN);
}

// ── Internal helpers ────────────────────────────────────────

function assertKdfNoneHasNoExtraFields(kdf: EnvelopeJson['kdf']): void {
  const raw = kdf as unknown as Record<string, unknown>;
  for (const field of [
    'version',
    'salt',
    'm_cost',
    't_cost',
    'p_cost',
    'length',
    'iterations',
  ]) {
    if (field in raw) {
      throw new Error(`kdf.name=none must not include ${field}`);
    }
  }
}

function parseAndValidateArgon2idKdf(kdf: KdfArgon2id): Uint8Array {
  if (kdf.version !== ARGON2_VERSION)
    throw new Error(`kdf.version must be ${ARGON2_VERSION}`);

  if (!Number.isInteger(kdf.m_cost))
    throw new Error('kdf.m_cost must be an integer');
  if (!Number.isInteger(kdf.t_cost))
    throw new Error('kdf.t_cost must be an integer');
  if (!Number.isInteger(kdf.p_cost))
    throw new Error('kdf.p_cost must be an integer');

  if (kdf.m_cost < ARGON2_M_COST_MIN || kdf.m_cost > ARGON2_M_COST_MAX) {
    throw new Error(
      `kdf.m_cost must be in range ${ARGON2_M_COST_MIN}..${ARGON2_M_COST_MAX}`,
    );
  }
  if (kdf.t_cost < ARGON2_T_COST_MIN || kdf.t_cost > ARGON2_T_COST_MAX) {
    throw new Error(
      `kdf.t_cost must be in range ${ARGON2_T_COST_MIN}..${ARGON2_T_COST_MAX}`,
    );
  }
  if (kdf.p_cost < ARGON2_P_COST_MIN || kdf.p_cost > ARGON2_P_COST_MAX) {
    throw new Error(
      `kdf.p_cost must be in range ${ARGON2_P_COST_MIN}..${ARGON2_P_COST_MAX}`,
    );
  }

  if (kdf.m_cost * kdf.t_cost > ARGON2_M_COST_T_COST_PRODUCT_MAX) {
    throw new Error(
      `kdf.m_cost * kdf.t_cost must be <= ${ARGON2_M_COST_T_COST_PRODUCT_MAX}`,
    );
  }

  if (kdf.length !== PASS_KEY_LEN)
    throw new Error(`kdf.length must be ${PASS_KEY_LEN}`);

  const salt = base64urlDecode(kdf.salt);
  if (salt.length < KDF_SALT_LEN) {
    throw new Error(`kdf.salt must be at least ${KDF_SALT_LEN} bytes`);
  }

  return salt;
}

async function hkdfDerive(
  ikm: Uint8Array,
  salt: Uint8Array,
  info: string,
  length: number,
): Promise<Uint8Array> {
  const key = await crypto.subtle.importKey('raw', buf(ikm), 'HKDF', false, [
    'deriveBits',
  ]);
  const bits = await crypto.subtle.deriveBits(
    {
      name: 'HKDF',
      hash: 'SHA-256',
      salt: buf(salt),
      info: buf(utf8Encode(info)),
    },
    key,
    length * 8,
  );
  return new Uint8Array(bits);
}

function cryptoRng(b: Uint8Array): void {
  crypto.getRandomValues(b);
}

// ── Bridge exports: dispatch to native (Tauri) or web crypto ──

export async function seal(
  content: Uint8Array,
  meta: PayloadMeta,
  options?: SealOptions,
): Promise<SealResult> {
  if (isTauri()) return nativeSeal(content, meta, options?.passphrase);
  return webSeal(content, meta, options);
}

export async function open(
  envelope: EnvelopeJson,
  urlKey: Uint8Array,
  passphrase?: string,
): Promise<OpenResult> {
  if (isTauri()) return nativeOpen(envelope, urlKey, passphrase);
  return webOpen(envelope, urlKey, passphrase);
}

export async function deriveClaimToken(
  urlKey: Uint8Array,
): Promise<Uint8Array> {
  if (isTauri()) return nativeDeriveClaimToken(urlKey);
  return webDeriveClaimToken(urlKey);
}
